lib/pushmi_pullyu/aip/downloader.rb

```diff
@@ -0,0 +1,215 @@
+require 'rdf/n3'
+require 'pushmi_pullyu/aip'
+require 'pushmi_pullyu/aip/solr_fetcher'
+require 'pushmi_pullyu/aip/fedora_fetcher'
+
+# Download all of the metadata/datastreams and associated data
+# related to an object
+
+class PushmiPullyu::AIP::Downloader
+
+  FILENAME_PREDICATE = 'info:fedora/fedora-system:def/model#downloadFilename'.freeze
+
+  def initialize(noid)
+    @noid = noid
+    @fedora_fetcher = PushmiPullyu::AIP::FedoraFetcher.new(@noid)
+    PushmiPullyu::AIP.validate(noid)
+  end
+
+  def run
+    make_directories
+    PushmiPullyu.logger.info("#{@noid}: Retreiving data from Fedora ...")
+
+    [:main_object, :fixity, :content_datastream_metadata, :versions, :thumbnail,
+     :characterization, :fedora3foxml, :fedora3foxml_metadata].each do |item|
+      path_spec = aip_paths.send(item)
```

Comment:
Instead of using ruby`s metaprogramming `send`, openstruct supports the `[](name)` notation. So this can simply be `aip_paths[item]` 

http://ruby-doc.org/stdlib-2.0.0/libdoc/ostruct/rdoc/OpenStruct.html#5B-5D-method